### ISSUE TITLE
[$500] IOU - Original amount is copied when system message shows the latest amount for paid edited request

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#33816]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.20-2
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
Precondition: User A and B have no unsettled IOU
1. [User B] Navigate to a different conversation
2. [User A] Request money from User B
3. [User A] Change the requested amount
4. [User B] Navigate to the main chat with User A
5. [User B] Click Pay elsewhere without going to IOU report
6. [User B] Go to IOU report
7. [User B] Right click on the paid system message > Copy to clipboard
8. [User B] Paste the content

## Expected Result:
The copied amount is the same as the amount in the system message, which is the latest amount

## Actual Result:
The copied amount is the original requested amount, while the system message shows the latest amount

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/297d2fa6-bbe1-4976-bfa9-b8f07dbc952c

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01447abde9208e3a7c</li>
        <li>Upwork Job ID: 1741885146524594176</li>
        <li>Last Price Increase: 2024-01-22</li>
    </ul>
</details>

**Comment by brookbdt** [comment]
I have some possible workarounds!

(Client-side)
 -> ensure that the action dispatched to update the requested amount is correctly modifying the application's state.
 -> ensure that the component displaying the system message is reactively bound to the updated state.
 -> ensure that the messages concerning payment updates are correctly processed and that their payload correctly updates 
      the relevant state. ( Assuming websockets or polling )

(Server-side)
-> When User A updates the requested amount, the back-end should properly handle this update in the database.
-> Ensure that any API endpoint responsible for fetching the payment request data returns the latest state.
-> Verify robust transaction handling in the backend.
-> Use appropriate database locking mechanisms to prevent race conditions, especially if multiple updates can happen in quick succession

In conclusion, by focusing on robust state management, reactive data binding, and ensuring API and database consistency, the issue can be systematically identified and resolved. I am confident it's one of the things listed above!

**Comment by brookbdt** [comment]
Contributor details
Your Expensify account email: brookbdt@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~0139f2296710c2db01

**Comment by brookbdt** [comment]
I have some possible workarounds! (Posting again in case you didn't see the previous one since you didn't have my contributor details)

(Client-side)
-> ensure that the action dispatched to update the requested amount is correctly modifying the application's state.
-> ensure that the component displaying the system message is reactively bound to the updated state.
-> ensure that the messages concerning payment updates are correctly processed and that their payload correctly updates
the relevant state. ( Assuming websockets or polling )

(Server-side)
-> When User A updates the requested amount, the back-end should properly handle this update in the database.
-> Ensure that any API endpoint responsible for fetching the payment request data returns the latest state.
-> Verify robust transaction handling in the backend.
-> Use appropriate database locking mechanisms to prevent race conditions, especially if multiple updates can happen in quick succession

In conclusion, by focusing on robust state management, reactive data binding, and ensuring API and database consistency, the issue can be systematically identified and resolved. I am confident it's one of the things listed above!

**Comment by rrrshtt** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The original amount is copied in the clipboard when the system message shows a different amount.

### What is the root cause of that problem?
https://github.com/Expensify/App/blob/fe49e4f7fb9677a5834b53cb776f79fb20f893ce/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.js#L58-L63
In this line, we memoized "reportAction" for the context menu. This could be the root of the problem because we are retaining old information for the context menu.

### What changes do you think we should make in order to solve the problem?
We should ensure that the information in the context menu is up-to-date. For example, we can pass only the ID of the report instead of the full report to the context menu and retrieve information only when we press actual actions. Alternatively, we can update our memoized hook dependencies to make sure that the information is up-to-date.

### What alternative solutions did you explore? (Optional)

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Original amount is copied when system message shows the latest amount for paid edited request

### What is the root cause of that problem?
We are getting the amount from `originalMessage` here
https://github.com/Expensify/App/blob/689c22710e9a8f0457014630c8919ef344bacb0d/src/libs/ReportUtils.ts#L4103-L4104
But It has the non-updated the original value.
### What changes do you think we should make in order to solve the problem?

We can get calculate from via `getMoneyRequestReimbursableTotal` same as we did to display the report action
https://github.com/Expensify/App/blob/689c22710e9a8f0457014630c8919ef344bacb0d/src/libs/ReportUtils.ts#L2008-L2009
but in case the `iouReport` doesn't exist we will fallback to the `originalMessage` amount 
```
const {amount: amountFromMessage, currency} = originalMessage.IOUDetails ?? originalMessage;
        const amount = iouReport ? getMoneyRequestReimbursableTotal(iouReport as OnyxEntry<Report>) : amountFromMessage;
```
### What alternative solutions did you explore? (Optional)

**Comment by brunovjk** [comment]
@FitseTLT when the account B is about to pay, should we get an error feedback? To try again later, with the updated value?

**Comment by anmurali** [comment]
@eVoloshchak - you have some proposals to review!

**Comment by anmurali** [comment]
@eVoloshchak another bump here.

**Comment by eVoloshchak** [comment]
Looks like this might have been resolved, I'm unable to reproduce the issue
@rrrshtt, @FitseTLT, could you double-check if this is still reproducible for you? Thanks

**Comment by FitseTLT** [comment]
Yep Solved in BE. Can close it

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by anmurali** [comment]
Closing per https://github.com/Expensify/App/issues/33816#issuecomment-1896420104 and https://github.com/Expensify/App/issues/33816#issuecomment-1899316832
